### ISSUE TITLE
Union all select with different timezone get wrong result 

### ISSUE THREAD

**Issue by compasses** [ClickHouse/ClickHouse#63480]
Hi, don't why union all not work for different timezone:

```
WITH 1714551897000 AS time_utc
SELECT
    'UTC-12',
    toDateTime(time_utc, 'Etc/GMT+12')
UNION ALL
SELECT
    'UTC-11',
    toDateTime(time_utc, 'Pacific/Midway')
UNION ALL
SELECT
    'UTC-10',
    toDateTime(time_utc, 'Pacific/Honolulu')


   ┌─'UTC-12'─┬─toDateTime(time_utc, 'Etc/GMT+12')─┐
1. │ UTC-11   │                2106-02-06 18:28:15 │
   └──────────┴────────────────────────────────────┘
   ┌─'UTC-12'─┬─toDateTime(time_utc, 'Etc/GMT+12')─┐
2. │ UTC-12   │                2106-02-06 18:28:15 │
   └──────────┴────────────────────────────────────┘
   ┌─'UTC-12'─┬─toDateTime(time_utc, 'Etc/GMT+12')─┐
3. │ UTC-10   │                2106-02-06 18:28:15 │
   └──────────┴────────────────────────────────────┘

```

Plain select works well:
```
WITH 1714551897000 AS time_utc
SELECT
    'UTC-12',
    toDateTime(time_utc, 'Etc/GMT+12'),
    'UTC-11',
    toDateTime(time_utc, 'Pacific/Midway'),
    'UTC-10',
    toDateTime(time_utc, 'Pacific/Honolulu')

Query id: 98dc9828-2dd5-4a27-b251-7af4470a30c6

   ┌─'UTC-12'─┬─toDateTime(time_utc, 'Etc/GMT+12')─┬─'UTC-11'─┬─toDateTime(time_utc, 'Pacific/Midway')─┬─'UTC-10'─┬─toDateTime(time_utc, 'Pacific/Honolulu')─┐
1. │ UTC-12   │                2106-02-06 18:28:15 │ UTC-11   │                    2106-02-06 19:28:15 │ UTC-10   │                      2106-02-06 20:28:15 │
   └──────────┴────────────────────────────────────┴──────────┴────────────────────────────────────────┴──────────┴──────────────────────────────────────────┘

```

**Comment by den-crane** [comment]
This is a correct and expected result. 
Timezone is a property of a column not of a value. A column can have only one timezone.
The timezone from the first select is used.

**Comment by den-crane** [comment]
Also 1714551897000 it's not a DateTime, it's DateTime64.

https://fiddle.clickhouse.com/498612af-5194-410c-85d0-ea7e00aba2ee

https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions#fromunixtimestamp64milli

**Comment by compasses** [comment]
> This is a correct and expected result. Timezone is a property of a column not of a value. A column can have only one timezone. The timezone from the first select is used.

Hi, thanks, but why use literal time string it's also right:
```
WITH '2024-05-01 16:24:57' AS date_value
SELECT
    'UTC-12',
    toDateTime(date_value, 'Etc/GMT+12')
UNION ALL
SELECT
    'UTC-11',
    toDateTime(date_value, 'Pacific/Midway')
UNION ALL
SELECT
    'UTC-10',
    toDateTime(date_value, 'Pacific/Honolulu')

Query id: 891b449f-c9b8-42de-929e-4990bf857c71

   ┌─'UTC-12'─┬─toDateTime(date_value, 'Etc/GMT+12')─┐
1. │ UTC-12   │                  2024-05-01 16:24:57 │
   └──────────┴──────────────────────────────────────┘
   ┌─'UTC-12'─┬─toDateTime(date_value, 'Etc/GMT+12')─┐
2. │ UTC-11   │                  2024-05-01 15:24:57 │
   └──────────┴──────────────────────────────────────┘
   ┌─'UTC-12'─┬─toDateTime(date_value, 'Etc/GMT+12')─┐
3. │ UTC-10   │                  2024-05-01 14:24:57 │
   └──────────┴──────────────────────────────────────┘

```

**Comment by den-crane** [comment]
>Hi, thanks, but why use literal time string it's also right:

Because TZ is property of datetime type and its value is numeric (timestamp - number of ticks since the beginning of the epoch in UTC). TZ is used during the parsing of strings and during the rendering numeric values back to strings.

So in case of `toDateTime('2024-05-01 16:24:57', 'Pacific/Midway')` -> string form `'2024-05-01 16:24:57'` is converted using `Pacific/Midway` and the result is type `DateTime('Pacific/Midway')`.
And in case of `toDateTime(1714551897, 'Pacific/Midway')` -> numeric value is placed into a datetime value without using TZ with the type `DateTime('Pacific/Midway')`. 
In other words Clickhouse treats numbers 1714551897 as a fully prepared timestamp in UTC TZ. And strings '2024-05-01 16:24:57' as not prepared value which should be converted into a timestamp using TZ.

**Comment by compasses** [comment]
Got you, thanks.
